### PR TITLE
Newsletters: add 274 (2023-10-25)

### PR DESCRIPTION
**Pull request by harding**
- [x] Bitcoin StackExchange @bitschmidty 
- [x] Other stuff @harding

### CODE REVIEW

**Comment by harding** [comment]
Updated with the rest of this week's news content and my own self-edits.  I'll add a few items to the notable code changes section tomorrow (Tuesday) but I'm bumping some of the changes we would normally cover to next week.

Sorry for the lateness.

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+
+  _Replacement cycling_ is a technique that allows targeted removal of a
+  transaction from node mempools.  To take a simple example, Mallory
+  broadcasts a _precursor transaction_ that stays unconfirmed.  She then
+  spends one of the outputs from the precursor in a _removable
+  transaction_.  Later, she initiates a [replacement][topic rbf] of the
+  precursor transaction, which triggers a removal of the removable
+  transaction.
+
+  When used by Mallory on transactions she exclusively controls, there's
```
Comment (glozow) [inline _posts/en/newsletters/2023-10-25-newsletter.md:36]: nitty nit
```suggestion
  When used by Mallory on transactions she exclusively controls, there is
```

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+      to an arbitrary address if she first spends her funds in a
+      [vault][topic vaults] notification transaction and waits for some
+      number of blocks to allow a freeze transaction to block the spend.
+      However, if the arbitrary address includes an `OP_SUCCESSx`
+      opcode, anyone will be able to steal her money.  Russell suggests
+      two possible solutions to this problem.
+
+  The research received some discussion and Russell indicated that he is
+  working on a follow-up post related to introspection of output
+  amounts..
```
Comment (bitschmidty) [inline _posts/en/newsletters/2023-10-25-newsletter.md:383]:
```suggestion
  amounts.
```

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+  transaction from node mempools.  To take a simple example, Mallory
+  broadcasts a _precursor transaction_ that stays unconfirmed.  She then
+  spends one of the outputs from the precursor in a _removable
+  transaction_.  Later, she initiates a [replacement][] of precursor
+  transaction, which triggers a removal of the removable transaction.
+
+  When used by Mallory on transactions she exclusively controls, there's
+  no significant problem with replacement cycling.  Except in cases where
+  transactions get evicted from mempools for normal reasons, the BIP125
+  rules that Mallory's replacement must follew ensure that miners will
```
Comment (murchandamus) [inline _posts/en/newsletters/2023-10-25-newsletter.md]:
```suggestion
  rules that Mallory's replacement must follow ensure that miners will
```

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+  transaction fees paid to close and settle the Bob-MalloryB channel).
+
+  Alternatively, when Bob closes the channel and attempts to spend the
+  forwarded payment back to himself, MalloryB can replace his spend with a
+  spend of her own containing the preimage.  If that transaction confirmed
+  promptly, Bob would learn the preimage and be able to claim the
+  forwarded payment from MalloryA, making Bob happy.
+
+  However, if MalloryB replaces Bob's spend and a spend of her own that
+  contains the preimage but is also a removable tranaction, and she
```
Comment (murchandamus) [inline _posts/en/newsletters/2023-10-25-newsletter.md]:
```suggestion
  contains the preimage but is also a removable transaction, and she
```

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+      public LN forwarding nodes currently try to hide this.  Now Core
+      Lightning, Eclair, LDK, and LND will all rebroadcast more
+      frequently.
+
+      After each time Bob rebroadcasts, Mallory can use the same
+      technique to replace his transaction again.  However, the BIP125
+      replacement rules will require Mallory to pay additional
+      transaction fees for each of her replacements and removals,
+      meaning each rebroadcast by Bob lowers the profitability to
+      Mallory of a successful attack.
```
Comment (murchandamus) [inline _posts/en/newsletters/2023-10-25-newsletter.md:127]: Since Bob’s transaction always has the same feerate and that feerate must be uncompetitive for the attack to be viable, is it really necessary for Mallory to increasingly pay more fees? If e.g. Mallory operates a high-volume wallet, she could use a UTXO to do the replacement cycle and then use that input towards an actual transaction at a competitive feerate.

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+
+      This suggests a rough formula for the maximum amount of an HTLC
+      that a node should accept.  If the cost is the attacker will need
+      to pay for each round of replacement-and-removal is _x_, the
```
Comment (murchandamus) [inline _posts/en/newsletters/2023-10-25-newsletter.md]:
```suggestion
      that a node should accept.  If the cost that the attacker will need
      to pay for each round of replacement-and-removal is _x_, the
```

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+  Bitcoin-Dev and Lightning-Dev mailing list in response to the
+  disclosure of the replacement cycling attack.  Suggested responses
+  included the following:
+
+    - **Incrementing fees towards scorched earth:** Antoine Riard's
+      [paper][riard cycle paper] about the attack and mailing list posts
+      by [Ziggie][ziggie cycle] and [Matt Morehouse][morehouse cycle]
+      suggest that, instead of having the defender (e.g. Bob), just
+      rebroadcast his refund spend, he start broadcasting conflicting
+      alternative spends that pay ever increasing feerates as the
```
[the remark this answers is not in the document]
Comment (murchandamus) [reply _posts/en/newsletters/2023-10-25-newsletter.md:243]: Isn’t that just conjunctive mood?

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+  mitigations deployed so far by various LN implementations, and a
+  summary of mitigations and solutions proposed on the mailing list.
+
+  _Replacement cycling_ is a technique that allows targeted removal of a
+  transaction from node mempools.  To take a simple example, Mallory
+  broadcasts a _precursor transaction_ that stays unconfirmed.  She then
+  spends one of the outputs from the precursor in a _removable
+  transaction_.  Later, she initiates a [replacement][topic rbf] of the
+  precursor transaction, which triggers a removal of the removable
+  transaction.
```
[the remark this answers is not in the document]
Comment (harding) [reply _posts/en/newsletters/2023-10-25-newsletter.md:34]: @glozow 

Yeah, after writing this section and seeing a proposed fix that didn't seem to work, I [replied](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-October/022067.html) to someone on list to also say that I didn't think the precursor is necessary.

I'll see if I can revise this section.  In the meantime, I [asked a question](https://bitcoin.stackexchange.com/questions/120200/what-are-the-advantages-to-attackers-of-different-styles-of-replacement-cycle-at) about the tradeoffs of the different approaches.

Thanks for pointing this out!

File: _posts/en/newsletters/2023-10-25-newsletter.md (hunk 5 above)
Comment (harding) [reply _posts/en/newsletters/2023-10-25-newsletter.md:127]: > Since Bob’s transaction always has the same feerate and that feerate must be uncompetitive for the attack to be viable

By "uncompetitive", do you mean that Bob must be paying too low of a feerate?  That's not actually true.  Bob could pay a feerate of 1,000 sat/vb and Mallory would still be able to perform a replacement cycle, which might still be worth it if earns her a 1,000,000 sat HTLC.

Was it something I wrote that gave you the impression that Bob's feerate must be uncompetitive?

> If e.g. Mallory operates a high-volume wallet, she could use a UTXO to do the replacement cycle and then use that input towards an actual transaction at a competitive feerate.

That's correct, but it's also where the deployed mitigations and proposed mitigations come in.  If Bob rebroadcasts 5 times a block, Mallory will need to have an operation where shes send about 5 organic transactions per block in order to have a marginal cost of zero.  That's fairly high volume, and it scales almost linearly depending on how many HTLCs Mallory is attacking (e.g., if she had 10 HTLCs in the channel with Bob, she would need to be at the level of typically sending about 50 organic transactions per block).  The additional proposed mitigation of having Bob pay higher fees as the CLTV expiry delta eliminates this advantage.  If Bob pays 10x the next-block fee near the end of CLTV expiry delta, then Mallory's replacements will need to pay an absolute fee greater than that amount, which she can only spread among additional outputs in organic transactions up to the 100,000 vbyte limit.  After that, she'll need to overpay fees.

In other words, I *think* the mitigations quickly make it so that only the highest of high-frequency spenders could have a minimal marginal attack cost, and that would only be possible if they only attacked a few HTLCs at once.  If they attacked more than about 100 HTLCs simultaneously (which is just 1 HTLC per block with a CLTV expiry delta of 144), I think anyone would be guaranteed to pay an above-zero marginal cost.

File: _posts/en/newsletters/2023-10-25-newsletter.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (harding) [reply _posts/en/newsletters/2023-10-25-newsletter.md]: FYI, updated to use the simpler non-precursor scenario.   See https://github.com/bitcoinops/bitcoinops.github.io/pull/1365/commits/1f9d3eb61edacf2fabc83499a1c83a6b5b2f71ef (still working on other edits)

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 9]
```diff
@@ -0,0 +1,506 @@
...
+      Teinturier [replied][teinturier cycle], "I agree with Matt though
+      that more fundamental work most likely needs to happen at the
+      bitcoin layer to allow L2 protocols to be more robust against that
+      class of attacks."  Riard also [said][riard cycle3] something
+      similar, "a sustainable fix can [only] happen at the base-layer,
+      e.g adding a memory-intensive history of all seen transactions".
+
+    - **Presigned fee bumps:** Peter Todd [argued][todd cycle1] that,
+      "the correct way to do pre-signed transactions is to pre-sign
+      enough *different* transactions to cover all reasonable needs for
```
[the remark this answers is not in the document]
Comment (harding) [reply _posts/en/newsletters/2023-10-25-newsletter.md:284]: Single asterisk and single underscore both produce italics in Markdown, so I'd prefer to stick with the verbatin copy/paste.

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 10]
```diff
@@ -0,0 +1,506 @@
...
+      replacement cycles thread, Peter Todd [proposed][todd expire1]
+      several consensus changes to enable an `OP_EXPIRE` opcode will
+      make a transaction invalid for inclusion after a specified block
+      height if the transaction's script executes `OP_EXPIRE`.  This can
+      be used to make the Mallory's preimage condition of an HTLC only
+      usable before Bob's refund condition becomes spendable.  This
+      prevents Mallory from being able to replace Bob's refund spend,
+      making it impossible for Mallory to execute a replacement cycle
+      attack.  `OP_EXPIRE` may also address some [transaction pinning
+      attacks][topic transaction pinning] against HTLCs.
```
[the remark this answers is not in the document]
Comment (harding) [reply _posts/en/newsletters/2023-10-25-newsletter.md:315]: Good answer.  I upvoted it.  I didn't mention that concern here, in this very short summary, because Peter Todd took steps to address that concern in his proposal (to the same degree that coinbase transactions address that problem).

**Comment by harding** [comment]
All edits made; all fixmes fixed (but one added for @bitschmidty at publication time).  Reviewed B.SE section (thanks @bitschmidty !).  Huge thanks to @glozow @bitschmidty @LarryRuane @murchandamus for the detailed reviews on **the longest regular newsletter** we've ever published!  (by word count; regular == not year-in-review)

I haven't added the topic entries for this week or last week yet.

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 11]
```diff
@@ -0,0 +1,588 @@
...
+lang: en
+---
+This week's newsletter describes the replacement cycling attack against
+HTLCs used in LN and other systems, examines the mitigations deployed
+for the attack, and summarizes several proposals for additional
+mitigations.  Also described are a notable bug affecting a Bitcoin Core
+RPC, research into covenants with minimal changes to Bitcoin Script, and
+a proposed BIP for an `OP_CAT` opcode.  Also included is our regular
+monthly section with summaries of popular questions and answers from the
+Bitcoin StackExchange.
```
Comment (azuchi) [inline _posts/en/newsletters/2023-10-25-newsletter.md:17]:
```suggestion
Bitcoin Stack Exchange.
```

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 12]
```diff
@@ -0,0 +1,588 @@
...
+  rbf] to remove the one or more inputs of a multi-input transaction
+  from node mempools.  To take a simple example, one that [differs
+  slightly]({{bse}}120200) from Riard's original description, Mallory
+  broadcasts a transaction with two inputs, which spend outputs _A_ and
+  _B_.  She then replaces that transaction with an alternative
+  single-input version that only spends output _B_.  After that
+  replacement, input _A_---and any data included in it---has been
+  removed from any node mempools that processed the replacement.
+
+  Alhough it's not safe for a regular wallet to do that[^rbf-warning],
```
Comment (azuchi) [inline _posts/en/newsletters/2023-10-25-newsletter.md:49]:
```suggestion
  Although it's not safe for a regular wallet to do that[^rbf-warning],
```

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 13]
```diff
@@ -0,0 +1,588 @@
...
+  are affected; users who only use their channels to initiate and
+  receive payments are not affected.
+
+  We've organized our description of this story into three separate news
+  items: a description of the vulnerability (this item), a description
+  of the mitigations deployed so far by various LN implementations, and
+  a summary of additional mitigations and solutions proposed on the mailing list.
+
+  As background, it's possible to use [transaction replacement][topic
+  rbf] to remove the one or more inputs of a multi-input transaction
```
Comment (LarryRuane) [inline _posts/en/newsletters/2023-10-25-newsletter.md:40]:
```suggestion
  rbf] to remove one or more inputs of a multi-input transaction
```

**Comment by harding** [comment]
Made edits for @azuchi and @LarryRuane additional feedback (thanks!) and add topic entries for both this week and last week.

Thanks everyone!

File: _posts/en/newsletters/2023-10-25-newsletter.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (glozow) [reply _posts/en/newsletters/2023-10-25-newsletter.md]: Ah I didn't see that, thanks 👍

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 14]
```diff
@@ -0,0 +1,588 @@
...
+
+- [Bitcoin Core 24.2][] is a maintenance relaese mainly containing bug
+  fixes.  It is recommend for anyone still using 24.0 or 24.1 who is
+  unable or unwilling to upgrade to 25.1 at this time.
+
+<!-- FIXME: bitschmidty to test whether the 26.0 link works before
+publication.  If it does, please remove the "Verified test..." sentence.
+In either case, please also delete this comment.  (Thanks, Mike!) -->
+
+- [Bitcoin Core 26.0][] is a release candidate for the next major
```
Comment (glozow) [inline _posts/en/newsletters/2023-10-25-newsletter.md:495]: fyi @bitschmidty seems quite unlikely to have bins up before the newsletter is published.

Comment (vostrnad) [reply _posts/en/newsletters/2023-10-25-newsletter.md:495]: Also:
```suggestion
<!-- FIXME: bitschmidty to test whether the 26.0 link works before
publication.  If it does, please remove the "Verified test..." sentence.
In either case, please also delete this comment.  (Thanks, Mike!) -->

- [Bitcoin Core 26.0rc1][] is a release candidate for the next major
```

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 15]
```diff
@@ -0,0 +1,588 @@
...
+
+*New releases and release candidates for popular Bitcoin infrastructure
+projects.  Please consider upgrading to new releases or helping to test
+release candidates.*
+
+- [Bitcoin Core 25.1][] is a maintenance release mainly containing bug
+  fixes.  It is the current recommended version of Bitcoin Core.
+
+- [Bitcoin Core 24.2][] is a maintenance relaese mainly containing bug
+  fixes.  It is recommend for anyone still using 24.0 or 24.1 who is
```
Comment (vostrnad) [inline _posts/en/newsletters/2023-10-25-newsletter.md:488]:
```suggestion
  fixes.  It is recommended for anyone still using 24.0 or 24.1 who is
```

File: _posts/en/newsletters/2023-10-25-newsletter.md [hunk 16]
```diff
@@ -0,0 +1,588 @@
...
+  users.
+
+  If you're aware of
+  any software that uses the `hash_serialized_2` field, please notify
+  its authors about the problem and encourage them to read Jahr's email
+  about the changes being made for the next major version of Bitcoin
+  Core to address the bug.
+
+- **Research into generic covenants with minimal Script language changes:**
+  Rusty Russell [posted][russell scripts] to Bitcoin-Dev a link to some
```
Comment (azuchi) [inline _posts/en/newsletters/2023-10-25-newsletter.md:372]:
```suggestion
  Rusty Russell [posted][russell scripts] to Bitcoin-Dev mailing list a link to some
```

File: _topics/en/assumeutxo.md [hunk 17]
```diff
@@ -55,6 +55,9 @@ optech_mentions:
   - title: "Bitcoin Core #27596 adds assumedvalid snapshot chainstate and full validation sync in the background"
     url: /en/newsletters/2023/10/11/#bitcoin-core-27596
 
+  - title: Bitcoin Core bug found in compution of UTXO set hash
```
Comment (bitschmidty) [inline _topics/en/assumeutxo.md:58]:
```suggestion
  - title: Bitcoin Core bug found in computation of UTXO set hash
```

File: _posts/en/newsletters/2023-10-25-newsletter.md (hunk 16 above)
Comment (bitschmidty) [reply _posts/en/newsletters/2023-10-25-newsletter.md:372]: Took this suggestion with a slight addition

**Comment by bitschmidty** [comment]
Sorry for the flurry of activity here but I chose to take many of the suggestions piecemeal here in github. 

Took some outstanding suggestions from @azuchi and @vostrnad, ensured other comments and suggestions were addressed or resolved.

Pushed a small change to a topic spelling.

Made some updates wrt the Bitcoin Core 26.0 (rc1) release.

Added clarification to the RBF footnote which noted the source of the quote (Mastering Bitcoin) which I think preemptively answers a question from readers "which book??" while also avoiding self promotion since I put it in and Im not involved with the book's efforts.

Separated News273-related topics into their own commit. Squashed. 

Thanks @harding for the heroic efforts this week! 💪 🚀 
Thanks to @jirijakes @murchandamus @LarryRuane @azuchi @glozow @vostrnad for the feedback and review!
